hudi-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadata.java

```diff
@@ -210,38 +229,110 @@ public HoodieData<HoodieRecord<HoodieMetadataPayload>> getRecordsByKeyPrefixes(L
+      if (baseFileReader == null && logRecordScanner == null) {
+        // TODO: what do we do if both does not exist? should we throw an exception and let caller do the fallback ?
+        return Collections.emptyIterator();
+      }
+      boolean fullKeys = false;
 
-                boolean fullKeys = false;
+      Map<String, HoodieRecord<HoodieMetadataPayload>> logRecords =
+          readLogRecords(logRecordScanner, sortedKeyPrefixes, fullKeys, timings);
+      Map<String, HoodieRecord<HoodieMetadataPayload>> mergedRecords =
+          readFromBaseAndMergeWithLogRecords(baseFileReader, sortedKeyPrefixes, fullKeys, logRecords, timings, partitionName);
 
-                Map<String, HoodieRecord<HoodieMetadataPayload>> logRecords =
-                    readLogRecords(logRecordScanner, sortedKeyPrefixes, fullKeys, timings);
+      LOG.debug("Metadata read for {} keys took [baseFileRead, logMerge] {} ms", sortedKeyPrefixes.size(), timings);
+
+      return mergedRecords.values().iterator();
+    } catch (IOException ioe) {
+      throw new HoodieIOException("Error merging records from metadata table for  " + sortedKeyPrefixes.size() + " key : ", ioe);
+    } finally {
+      closeReader(readers);
+    }
+  }
 
-                Map<String, HoodieRecord<HoodieMetadataPayload>> mergedRecords =
-                    readFromBaseAndMergeWithLogRecords(baseFileReader, sortedKeyPrefixes, fullKeys, logRecords, timings, partitionName);
+  // TODO: add predicate support.
+  private Iterator<HoodieRecord<HoodieMetadataPayload>> getByKeyPrefixesWithFileGroupReader(FileSlice fileSlice,
+                                                                                            List<String> sortedKeyPrefixes,
+                                                                                            String partitionName) throws IOException {
+    Option<HoodieInstant> latestMetadataInstant =
+        metadataMetaClient.getActiveTimeline().filterCompletedInstants().lastInstant();
+    String latestMetadataInstantTime =
+        latestMetadataInstant.map(HoodieInstant::requestedTime).orElse(SOLO_COMMIT_TIMESTAMP);
+    Schema schema = HoodieAvroUtils.addMetadataFields(HoodieMetadataRecord.getClassSchema());
+    HoodieFileGroupReader fileGroupReader = getFileGroupReader(
+        metadataMetaClient.getTableConfig(),
+        metadataMetaClient.getBasePath().toString(),
+        latestMetadataInstantTime,
+        fileSlice,
+        schema,
+        schema,
+        Option.empty(),
+        metadataMetaClient,
+        new TypedProperties(),
+        Collections.emptyList()); // TODO: Any properties?
+    fileGroupReader.initRecordIterators();
+    ClosableIterator it = fileGroupReader.getClosableIterator();
+    return new HoodieRecordIterator(it, partitionName, sortedKeyPrefixes);
+  }
+
+  public static class HoodieRecordIterator implements Iterator<HoodieRecord<HoodieMetadataPayload>> {
+    private final ClosableIterator<IndexedRecord> baseIterator;
```

Comment:
can you help understand, why do we need to read it as IndexedRecord and then convert it back to Metadata payload. 

From what I see, in existing places. 
We read from FileGroupReader as InternalRow and then construct the record as below. 

HoodieSparkRecord record = new HoodieSparkRecord(recordKey, row, sparkSchema, false);

Can't we do something like that?

Reply:
We can support customized HoodieRecord iterator on FG reader based on: https://github.com/apache/hudi/pull/13314